pkg/addon/addon_test.go

```diff
@@ -1260,3 +1260,100 @@ func TestGenerateAnnotation(t *testing.T) {
 	assert.Equal(t, res[velaSystemRequirement], "")
 	assert.Equal(t, res[kubernetesSystemRequirement], ">=1.20.1")
 }
+
+func TestMergeAddonInstallArgs(t *testing.T) {
+	k8sClient := fake.NewClientBuilder().Build()
+	ctx := context.Background()
+
+	testcases := []struct {
+		name        string
+		legacyArgs  string
+		args        map[string]interface{}
+		mergedArgs  string
+		application string
+		err         error
+	}{
+		{
+			name:       "addon1",
+			legacyArgs: "{\"clusters\":[\"\"],\"imagePullSecrets\":[\"test-hub\"],\"repo\":\"hub.vela.com\",\"serviceType\":\"NodePort\"}",
+			args: map[string]interface{}{
+				"serviceType": "NodePort",
+			},
+			mergedArgs: "{\"clusters\":[\"\"],\"imagePullSecrets\":[\"test-hub\"],\"repo\":\"hub.vela.com\",\"serviceType\":\"NodePort\"}",
+		},
+		{
+			name:       "addon2",
+			legacyArgs: "{\"clusters\":[\"\"]}",
+			args: map[string]interface{}{
+				"repo":             "hub.vela.com",
+				"serviceType":      "NodePort",
+				"imagePullSecrets": []string{"test-hub"},
+			},
+			mergedArgs: "{\"clusters\":[\"\"],\"imagePullSecrets\":[\"test-hub\"],\"repo\":\"hub.vela.com\",\"serviceType\":\"NodePort\"}",
+		},
+		{
+			name:       "addon3",
+			legacyArgs: "{\"clusters\":[\"\"],\"imagePullSecrets\":[\"test-hub\"],\"repo\":\"hub.vela.com\",\"serviceType\":\"NodePort\"}",
+			args: map[string]interface{}{
+				"imagePullSecrets": []string{"test-hub-2"},
+			},
+			mergedArgs: "{\"clusters\":[\"\"],\"imagePullSecrets\":[\"test-hub-2\"],\"repo\":\"hub.vela.com\",\"serviceType\":\"NodePort\"}",
+		},
+		{
+			// merge nested parameters
+			name:       "addon4",
+			legacyArgs: "{\"clusters\":[\"\"],\"p1\":{\"p11\":\"p11-v1\",\"p12\":\"p12-v1\"}}",
+			args: map[string]interface{}{
+				"p1": map[string]interface{}{
+					"p12": "p12-v2",
+					"p13": "p13-v1",
+				},
+			},
+			mergedArgs: "{\"clusters\":[\"\"],\"p1\":{\"p11\":\"p11-v1\",\"p12\":\"p12-v2\",\"p13\":\"p13-v1\"}}",
+		},
+		{
+			// there is not legacyArgs
+			name:       "addon5",
+			legacyArgs: "",
+			args: map[string]interface{}{
+				"p1": map[string]interface{}{
+					"p12": "p12-v2",
+					"p13": "p13-v1",
+				},
+			},
+			mergedArgs: "{\"p1\":{\"p12\":\"p12-v2\",\"p13\":\"p13-v1\"}}",
+		},
+		{
+			// there is not new args
+			name:       "addon6",
+			legacyArgs: "{\"clusters\":[\"\"],\"p1\":{\"p11\":\"p11-v1\",\"p12\":\"p12-v1\"}}",
+			args:       nil,
+			mergedArgs: "{\"clusters\":[\"\"],\"p1\":{\"p11\":\"p11-v1\",\"p12\":\"p12-v1\"}}",
+		},
+	}
```

Comment:
please add a case to test the override scenario

Reply:
nested + override

Reply:
please refer to Line 1304.  p1.p11 keeps legacy value, p1.p12 will overwrite legacy value, p1.p13 is one new input parameter.

Reply:
sorry, I missed that case, thanks